doc/Data Plane Code Generation/ServiceDrivenEvolution.md

```diff
@@ -0,0 +1,177 @@
+{
+    Argument.AssertNotNull(requiredParam, nameof(requiredParam));
+
+    using var scope = ClientDiagnostics.CreateScope("ParamsClient.PutRequiredOptional");
+    scope.Start();
+    try
+    {
+        using HttpMessage message = CreatePutRequiredOptionalRequest(requiredParam, optionalParam, context);
+        return await _pipeline.ProcessMessageAsync(message, context).ConfigureAwait(false);
+    }
+    catch (Exception e)
+    {
+        scope.Failed(e);
+        throw;
+    }
+}
+```
+
+Generated code in a V2 client with a new optional query parameter:
+
+``` C#
+public virtual async Task<Response> PutRequiredOptionalAsync(string requiredParam, string optionalParam = null, string newParameter = null, RequestContext context = null)
+{
+    Argument.AssertNotNull(requiredParam, nameof(requiredParam));
+
+    using var scope = ClientDiagnostics.CreateScope("ParamsClient.PutRequiredOptional");
+    scope.Start();
+    try
+    {
+        using HttpMessage message = CreatePutRequiredOptionalRequest(requiredParam, optionalParam, newParameter, context);
+        return await _pipeline.ProcessMessageAsync(message, context).ConfigureAwait(false);
+    }
+    catch (Exception e)
+    {
+        scope.Failed(e);
+        throw;
+    }
+}
+```
+
+Since `context` is always the last parameter, V2 inserts the new optional parameter before `context` parameter and this could break code that works with V1 such as `client.PutRequiredOptionalAsync("requiredParam", "optionalParam", ErrorOptions.NoThrow)` (`ErrorOptions.NoThrow` was passed in as `context` parameter but now it's treated as `newParameter`). We can manually add an overload method in V2 which transforms the optional parameters in V1 to required parameters to make V2 backward compatible with V1:
+
+``` C#
+public virtual async Task<Response> PutRequiredOptionalAsync(string requiredParam, string optionalParam, RequestContext context)
+{
+    Argument.AssertNotNull(requiredParam, nameof(requiredParam));
+
+    using var scope = ClientDiagnostics.CreateScope("ParamsClient.PutRequiredOptional");
+    scope.Start();
+    try
+    {
+        using HttpMessage message = CreatePutRequiredOptionalRequest(requiredParam, optionalParam, null, context);
+        return await _pipeline.ProcessMessageAsync(message, context).ConfigureAwait(false);
+    }
+    catch (Exception e)
+    {
+        scope.Failed(e);
+        throw;
+    }
+}
+```
+
+Now calling `client.PutRequiredOptionalAsync("requiredParam", "optionalParam", ErrorOptions.NoThrow)` with V2 will invoke the manual method we added.
+
+## A new body type is added
+
+Generated code in a V1 client with only `application/json` as the content type:
+
+``` C#
+public virtual async Task<Response> PostParametersAsync(RequestContent content, RequestContext context = null)
+{
+    Argument.AssertNotNull(content, nameof(content));
```

Comment:
We shouldn't be validating request bodies in our clients: https://azure.github.io/azure-sdk/dotnet_introduction.html#parameter-validation

Reply:
I created an issue here: https://github.com/Azure/autorest.csharp/issues/2155

Reply:
This looks by design, see https://github.com/Azure/autorest.csharp/issues/2155#issuecomment-1101111229